File: examples/model-a/manual-multi-step/initiator.py
"""Model A — Manual Multi-Step: initiator chains two agents sequentially.

The initiator manually orchestrates a 2-step workflow without ScenarioBundle:
  Step 1: compliance-checker validates the change
  Step 2: risk-assessment-agent scores the risk (receives step 1 result)

Run:
    # Terminal 1 — start both agents
    AXME_API_KEY=<compliance-key> AXME_AGENT_ADDRESS=compliance-checker-agent \
      python examples/delivery/stream/agent.py &
    AXME_API_KEY=<risk-key> AXME_AGENT_ADDRESS=risk-assessment-agent \
      python examples/internal/notification/agent.py &

    # Terminal 2 — run the initiator
    AXME_API_KEY=<workspace-api-key> \
    AXME_AGENT_1=agent://org/ws/compliance-checker-agent \
    AXME_AGENT_2=agent://org/ws/risk-assessment-agent \
      python examples/model-a/manual-multi-step/initiator.py
"""
from __future__ import annotations

import logging
import os
import sys

from axme import AxmeClient, AxmeClientConfig

logging.basicConfig(level=logging.INFO, format="%(asctime)s  [%(levelname)s]  %(message)s", datefmt="%H:%M:%S")
log = logging.getLogger("manual-multi-step")

AXME_BASE_URL = os.environ.get("AXME_BASE_URL", "https://api.cloud.axme.ai")
AXME_API_KEY  = os.environ.get("AXME_API_KEY", "")
AGENT_1       = os.environ.get("AXME_AGENT_1", "")
AGENT_2       = os.environ.get("AXME_AGENT_2", "")

if not AXME_API_KEY:
    sys.exit("AXME_API_KEY is required")
if not AGENT_1 or not AGENT_2:
    sys.exit("AXME_AGENT_1 and AXME_AGENT_2 are required (full canonical agent:// addresses)")


def _wait_for_done(client: AxmeClient, intent_id: str, timeout: int = 60) -> str:
    """Wait until the intent reaches a done state (IN_PROGRESS or terminal).

    In Model A (no agent-core), resume_intent sets IN_PROGRESS which is the
    agent's "done" signal.
    """
    for event in client.observe(intent_id, timeout_seconds=timeout):
        status = event.get("status", "")
        if status in ("COMPLETED", "IN_PROGRESS", "FAILED", "CANCELED", "TIMED_OUT"):
            return status
    return "UNKNOWN"


def main() -> None:
    client = AxmeClient(AxmeClientConfig(api_key=AXME_API_KEY, base_url=AXME_BASE_URL))

    change_payload = {
        "change_id": "CHG-MULTI-STEP-001",
        "service": "api-gateway",
        "version": "5.0.0",
        "environment": "staging",
        "change_type": "config_update",
        "risk_level": "medium",
        "risk_threshold": 30,
    }

    # ── Step 1: Compliance Check ──────────────────────────────────────
    log.info("STEP 1: compliance check → %s", AGENT_1)
    step1_id = client.send_intent(
        {
            "intent_type": "intent.compliance.check.v1",
            "from_agent": "initiator://manual-multi-step",
            "to_agent": AGENT_1,
            "payload": change_payload,
        },
    )
    log.info("step 1 intent: %s — waiting...", step1_id)
    step1_status = _wait_for_done(client, step1_id)
    log.info("step 1 → %s", step1_status)

    if step1_status not in ("COMPLETED", "IN_PROGRESS"):
        log.error("step 1 failed (%s) — aborting pipeline", step1_status)
        return

    # ── Step 2: Risk Assessment ───────────────────────────────────────
    log.info("STEP 2: risk assessment → %s", AGENT_2)
    step2_id = client.send_intent(
        {
            "intent_type": "intent.risk.assessment.v1",
            "from_agent": "initiator://manual-multi-step",
            "to_agent": AGENT_2,
            "payload": {
                **change_payload,
                "compliance_result": "passed",
                "compliance_intent_id": step1_id,
            },
        },
    )
    log.info("step 2 intent: %s — waiting...", step2_id)
    step2_status = _wait_for_done(client, step2_id)
    log.info("step 2 → %s", step2_status)

    # ── Summary ───────────────────────────────────────────────────────
    log.info("")
    log.info("=== Manual Multi-Step Pipeline ===")
    log.info("  Step 1 (compliance): %s  id=%s", step1_status, step1_id)
    log.info("  Step 2 (risk):       %s  id=%s", step2_status, step2_id)
    if step1_status in ("COMPLETED", "IN_PROGRESS") and step2_status in ("COMPLETED", "IN_PROGRESS"):
        log.info("  Result: ALL STEPS PASSED")
    else:
        log.info("  Result: PIPELINE FAILED")


if __name__ == "__main__":
    main()

File: examples/model-a/manual-multi-step/typescript/initiator.ts
/**
 * Model A — Manual Multi-Step: chain two agents sequentially (TypeScript).
 *
 * Run:
 *   AXME_API_KEY=<workspace-key> \
 *   AXME_AGENT_1=agent://org/ws/compliance-checker-agent \
 *   AXME_AGENT_2=agent://org/ws/risk-assessment-agent \
 *     npx tsx examples/model-a/manual-multi-step/initiator.ts
 */
import { AxmeClient } from "@axme/axme";

const AXME_BASE_URL = process.env.AXME_BASE_URL ?? "https://api.cloud.axme.ai";
const AXME_API_KEY = process.env.AXME_API_KEY ?? "";
const AGENT_1 = process.env.AXME_AGENT_1 ?? "";
const AGENT_2 = process.env.AXME_AGENT_2 ?? "";

if (!AXME_API_KEY) { console.error("AXME_API_KEY is required"); process.exit(1); }
if (!AGENT_1 || !AGENT_2) { console.error("AXME_AGENT_1 and AXME_AGENT_2 are required"); process.exit(1); }

const client = new AxmeClient({ apiKey: AXME_API_KEY, baseUrl: AXME_BASE_URL });

async function waitForDone(intentId: string, timeout = 60_000): Promise<string> {
  for await (const event of client.observe(intentId, { timeoutMs: timeout })) {
    const status = String(event.status ?? "");
    if (["COMPLETED", "IN_PROGRESS", "FAILED", "CANCELED", "TIMED_OUT"].includes(status)) return status;
  }
  return "UNKNOWN";
}

const changePayload = {
  change_id: "CHG-MULTI-STEP-001", service: "api-gateway", version: "5.0.0",
  environment: "staging", change_type: "config_update", risk_level: "medium", risk_threshold: 30,
};

// Step 1: Compliance Check
console.log(`STEP 1: compliance check → ${AGENT_1}`);
const step1Id = await client.sendIntent({
  intent_type: "intent.compliance.check.v1",
  from_agent: "initiator://manual-multi-step", to_agent: AGENT_1, payload: changePayload,
});
console.log(`step 1 intent: ${step1Id} — waiting...`);
const step1Status = await waitForDone(step1Id);
console.log(`step 1 → ${step1Status}`);

if (!["COMPLETED", "IN_PROGRESS"].includes(step1Status)) {
  console.error(`step 1 failed (${step1Status}) — aborting pipeline`);
  process.exit(1);
}

// Step 2: Risk Assessment
console.log(`STEP 2: risk assessment → ${AGENT_2}`);
const step2Id = await client.sendIntent({
  intent_type: "intent.risk.assessment.v1",
  from_agent: "initiator://manual-multi-step", to_agent: AGENT_2,
  payload: { ...changePayload, compliance_result: "passed", compliance_intent_id: step1Id },
});
console.log(`step 2 intent: ${step2Id} — waiting...`);
const step2Status = await waitForDone(step2Id);
console.log(`step 2 → ${step2Status}`);

// Summary
console.log("");
console.log("=== Manual Multi-Step Pipeline ===");
console.log(`  Step 1 (compliance): ${step1Status}  id=${step1Id}`);
console.log(`  Step 2 (risk):       ${step2Status}  id=${step2Id}`);
if (["COMPLETED", "IN_PROGRESS"].includes(step1Status) && ["COMPLETED", "IN_PROGRESS"].includes(step2Status)) {
  console.log("  Result: ALL STEPS PASSED");
} else {
  console.log("  Result: PIPELINE FAILED");
}
